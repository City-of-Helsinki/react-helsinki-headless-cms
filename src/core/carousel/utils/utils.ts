/**
 * @param arr - any array.
 * @param len - the array length.
 * @return chunks array of initial array (grouped elements).
 */
export function splitArrayIntoChunksOfLen(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  arr: Array<React.ReactElement<any>>,
  len: number,
) {
  const chunks = [];
  let i = 0;
  const n = arr.length;
  while (i < n) {
    chunks.push(arr.slice(i, (i += len)));
  }
  return chunks;
}

/**
 * @param index - item set index.
 * @param itemSetId - item set (element of the array).
 * @return an unique key based on input data.
 */
export const getItemSetKey = (index: number, itemSetId?: string) =>
  `itemSet-${itemSetId ?? Math.random()}-${index}`;

/**
 * Create an unique key for item set.
 * @param index - item index.
 * @param itemId - item (element of the array). Default 'itemSet-item'
 * @param itemSetPrefix - item prefix (element of the array).
 * @return an unique key based on input data.
 */
export const getItemSetItemKey = (
  index: number,
  itemId?: string,
  itemSetPrefix = 'itemSet-item',
) => `${itemSetPrefix}-${itemId ?? Math.random()}-${index}`;

/**
 * @param item - item (element of the array).
 * @param index - item index.
 * @return an unique key for the dot (slide counter) based on input data.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const getSlideDotKey = (item: any, index: number) =>
  `slide-dot-${item?.id ?? Math.random()}-${index}`;

export const getLoadMoreKey = () => `carousel-loadmore-${Math.random()}`;
