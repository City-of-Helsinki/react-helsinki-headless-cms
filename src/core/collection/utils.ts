import isPast from 'date-fns/isPast';

import type { EventType } from '../../common/eventsService/types';
import normalizeKeys from '../../linkedEvents/utils/normalizeKeys';
import { LINKEDEVENTS_DATE_NOW } from './constants';

/**
 * @param {ids} ids - the venue ids in number format.
 * @return {Array} - Returns array of tprek formatted string ids.
 */
export function getVenueIds(ids: number[]): string[] {
  return ids.map((id) => `tprek:${id}`);
}

/**
 * @param {event} - event.
 * @return {boolean} - Returns true if events' end time in past.
 */
export const isEventEndTimeInPast = (event: EventType): boolean =>
  !!event?.endTime && isPast(new Date(event.endTime));

/**
 * @param {event} - event.
 * @return {boolean} - Returns true if events' end time in null and start time in past.
 */
export const isEventEndTimeNullAndStartTimeInPast = (
  event: EventType,
): boolean =>
  !!event?.startTime &&
  event?.endTime === null &&
  isPast(new Date(event.startTime));

/**
 * @param {event} - event.
 * @return {boolean} - Returns true if event is closed.
 */
export const isEventClosed = (event: EventType): boolean =>
  isEventEndTimeInPast(event) || isEventEndTimeNullAndStartTimeInPast(event);

/**
 * @param {params} - Get params.
 * @return {object} - Returns normalized values.
 */
export const normalizeParamsValues = (params: Record<string, string>) => {
  const normalizedParams = { ...normalizeKeys(params) };

  // Fix for course event type lower case
  if (normalizedParams.eventType) {
    normalizedParams.eventType =
      normalizedParams.eventType.charAt(0).toUpperCase() +
      normalizedParams.eventType.slice(1);
  }

  // Fix boolean types
  Object.keys(normalizedParams).forEach((param) => {
    if (normalizedParams[param] === 'true') {
      normalizedParams[param] = true;
    } else if (normalizedParams[param] === 'false') {
      normalizedParams[param] = false;
    }
  });

  return normalizedParams;
};

/**
 * @param {dateString} - Get date string.
 * @return {boolean} - Returns true if the date is valid
 */
export function isValidDate(dateString: string) {
  return !Number.isNaN(Date.parse(dateString));
}

/**
 * @param {start} start - Get start date.
 * @return {boolean} - Returns current date as string if the initial was in past.
 */
export const convertDateStringInPastToNow = (start: string): string => {
  const startDate = new Date(start);
  const now = new Date();

  if (isValidDate(start) && startDate <= now) {
    return LINKEDEVENTS_DATE_NOW;
  }
  return start;
};
