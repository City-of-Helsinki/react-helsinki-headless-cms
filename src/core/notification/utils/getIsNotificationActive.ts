import isFuture from 'date-fns/isFuture';
import isPast from 'date-fns/isPast';
import isSameDay from 'date-fns/isSameDay';

import { NotificationType } from '../../../common/headlessService/types';

export default function isNotificationActive(
  notification?: NotificationType | null,
) {
  const { startDate, endDate, title, content } = notification ?? {};
  const isNotificationContentMissing = !content && !title;

  if (!notification || isNotificationContentMissing) {
    return false;
  }

  const isSameDayOrFuture = (date: Date) =>
    isSameDay(new Date(), date) || isFuture(date);

  return [
    // both start and end dates are defined
    () =>
      startDate &&
      endDate &&
      isPast(new Date(startDate)) &&
      isSameDayOrFuture(new Date(endDate)),
    // only end date defined
    () => !startDate && endDate && isSameDayOrFuture(new Date(endDate)),
    // only start date defined
    () => startDate && !endDate && isPast(new Date(startDate)),
    // neither start or end date are defined
    () => !startDate && !endDate,
  ].some((f) => !!f());
}
