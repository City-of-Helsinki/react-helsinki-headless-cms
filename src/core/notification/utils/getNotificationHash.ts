import hash from "../../../common/utils/hash";
import { NotificationType } from "../../../common/headlessService/types";

// The notification is a singleton in the CMS. Instead of there being multiple
// notifications, there ever only is one which is edited according to current
// needs.

// Because this notification comes with a dismissal option, we have to be able
// to distinct different versions of the notification, so that we can safely
// determine that the user has previously dismissed this notification (and
// received) the urgent information that's within the notification.
export default function getNotificationHash(
  notification?: NotificationType | null
) {
  // Notification fields used to build string for unique hash
  const stringKeys = [
    "title",
    "content",
    "linkText",
    "linkUrl",
    "level",
    "endDate",
    "startDate",
  ] as const;
  const combinedString = stringKeys.reduce(
    (acc, current) => acc + (notification?.[current] ?? ""),
    ""
  );

  return hash(combinedString);
}
