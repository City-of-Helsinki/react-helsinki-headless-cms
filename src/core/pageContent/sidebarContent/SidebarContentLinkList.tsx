import { IconArrowRight, IconLinkExternal } from 'hds-react';
import React from 'react';

import ExternalLink from '../../../common/components/externalLink/ExternalLink';
import HtmlToReact from '../../../common/components/htmlToReact/HtmlToReact';
import List from '../../../common/components/list/List';
import { LayoutLinkList } from '../../../common/headlessService/types';
import styles from './sidebarContentLinkList.module.scss';

type Link = {
  target?: string | null;
  title?: string | null;
  url?: string | null;
};

type SidebarContentLinkListProps = Omit<LayoutLinkList, '__typename'>;

export default function SidebarContentLinkList({
  anchor,
  title,
  description,
  links,
}: SidebarContentLinkListProps) {
  return (
    <div id={anchor || undefined}>
      <h2>{title}</h2>
      {description && (
        <div className={styles.descriptionContainer}>
          <HtmlToReact>{description}</HtmlToReact>
        </div>
      )}
      <List
        variant="spacing-s"
        items={links
          ?.filter((item): item is Link => Boolean(item))
          .map((link) =>
            link.target === '_blank' ? (
              <ExternalLink
                key={link.title}
                href={link.url || '#'}
                className={styles.link}
                iconLeft={<IconLinkExternal aria-hidden="true" />}
              >
                {link.title}
              </ExternalLink>
            ) : (
              <a
                key={link.title}
                href={link.url || '#'}
                className={styles.link}
              >
                <IconArrowRight aria-hidden="true" />
                {link.title}
              </a>
            ),
          )}
      />
    </div>
  );
}
