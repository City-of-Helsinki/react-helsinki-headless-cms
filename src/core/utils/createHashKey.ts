const MAX_KEY_LENGTH = 200;
export default function createHashKey(source: string | object) {
  const data = typeof source === 'string' ? source : JSON.stringify(source);
  return btoa(unescape(encodeURIComponent(data))).substring(0, MAX_KEY_LENGTH);
}
