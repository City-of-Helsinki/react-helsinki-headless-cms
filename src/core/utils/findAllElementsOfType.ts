import { getChildrenByType } from './getChildrenByType';
import { recursiveMap } from './recursiveMap';

/**
 * Find all elements of defined types recursively from a ReactNode.
 * @returns a list of elements of specified types that could be found from inside the ReactNode.
 */
export const findAllElementsOfType = (
  children: React.ReactNode,
  types: string[],
) => {
  const elementsOfType = [];
  recursiveMap(children, (child) => {
    elementsOfType.push(getChildrenByType(child, types));
    return child;
  });
  return elementsOfType.flat();
};
