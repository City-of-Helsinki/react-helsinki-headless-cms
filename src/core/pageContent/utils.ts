import { EventType } from '../../common/eventsService/types';
import {
  ArticleType,
  PageModule,
  PageType,
} from '../../common/headlessService/types';
import {
  getElementTextContent,
  isArticleType,
  isEventSearch,
  isEventSearchCarousel,
  isEventSelected,
  isEventSelectedCarousel,
  isEventType,
  isLayoutArticle,
  isLayoutArticleCarousel,
  isLayoutPage,
  isLayoutPageCarousel,
  isPageType,
} from '../../common/headlessService/utils';
import { CardProps } from '../card/Card';
import { type CollectionProps } from '../collection/Collection';
import {
  CollectionType,
  EventSearchCollectionType,
  EventSelectionCollectionType,
  GeneralCollectionType,
} from '../collection/types';

export function getCollections(
  pageModules: PageModule[],
  isEventSearchCollectionsEnables = false,
): CollectionType[] {
  return pageModules?.reduce((collections, module, index) => {
    const commonFields: CollectionType = {
      id: index.toString(),
      title: 'title' in module ? module.title : null,
      description: 'description' in module ? module.description : null,
      items: [],
      // eslint-disable-next-line no-underscore-dangle
      __typename: module.__typename,
    };

    if (isLayoutArticle(module) || isLayoutArticleCarousel(module)) {
      collections.push({
        ...commonFields,
        items: module.articles,
        showAllUrl: module.showAllLink,
      });
    }
    if (isLayoutPage(module) || isLayoutPageCarousel(module)) {
      collections.push({
        ...commonFields,
        items: module.pages,
      });
    }
    if (isEventSearchCollectionsEnables) {
      if (isEventSearch(module) || isEventSearchCarousel(module)) {
        collections.push({
          ...commonFields,
          url: module.url,
        } as EventSearchCollectionType);
      }
      if (isEventSelected(module) || isEventSelectedCarousel(module)) {
        collections.push({
          ...commonFields,
          events: module.events,
        } as EventSelectionCollectionType);
      }
    }
    return collections;
  }, []);
}

export function getArticlePageCardProps(
  item: ArticleType | PageType,
): CardProps {
  return {
    id: item.id,
    title: item.title,
    url: item.link,
    imageUrl: item.featuredImage?.node?.mediaItemUrl,
    ariaLabel: item.title,
    text: getElementTextContent((item.lead || item.content) ?? ''),
    hasLink: true,
    withBorder: true,
    withShadow: false,
    clampText: true,
    direction: 'responsive' as CardProps['direction'],
    openLinkInNewTab: false,
  };
}

export function getEventCardProps(item: EventType, locale = 'fi'): CardProps {
  const image = item.images.length > 0 ? item.images[0] : null;
  return {
    id: item.id,
    title: item.name[locale],
    url: item.internalId,
    imageUrl: image?.url,
    ariaLabel: item.name[locale],
    hasLink: true,
    withBorder: true,
    withShadow: false,
    clampText: true,
    direction: 'responsive' as CardProps['direction'],
    openLinkInNewTab: false,
  };
}

export function getCollectionCards(
  collection: GeneralCollectionType,
  locale = 'fi',
): CardProps[] {
  return collection.items.reduce((result: CardProps[], item) => {
    if (isPageType(item) || isArticleType(item))
      result.push(getArticlePageCardProps(item));
    else if (isEventType(item)) result.push(getEventCardProps(item, locale));

    return result;
  }, []);
}

export function getCollectionUIType(
  collection: CollectionType,
): CollectionProps['type'] {
  // eslint-disable-next-line no-underscore-dangle
  return collection.__typename.includes('Carousel') ? 'carousel' : 'grid';
}
