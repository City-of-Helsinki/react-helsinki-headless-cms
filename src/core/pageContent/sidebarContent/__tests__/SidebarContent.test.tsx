import React from 'react';

import SidebarContentLinkList from '../SidebarContentLinkList';
import { render } from '../../../../common/utils/testingLibrary';

describe('SidebarContent containing link lists', () => {
  const baseLinkItem = {
    title: 'title',
    url: '#',
  };

  const nullishValues = [undefined, null, ''];

  it('renders a link list when a valid item is included in the links', () => {
    const links = [
      { url: '#', title: 'link1' },
      { url: 'https://www.hel.fi', title: 'link2' },
    ];
    const { container } = render(
      <SidebarContentLinkList
        title="Renders links into a link list"
        links={links}
      />,
    );
    expect(
      container
        .getElementsByClassName('listContainer')[0]
        .getElementsByTagName('ul')[0].children,
    ).toHaveLength(links.length);
    expect(container).toMatchSnapshot();
  });

  it.each(['url', 'title'])(
    'skips the link items that does not have %s which is a mandatory field',
    (mandatoryField) => {
      nullishValues.forEach((nullishValue) => {
        const linkItem = {
          ...baseLinkItem,
          [mandatoryField]: nullishValue,
        };
        const { container } = render(
          <SidebarContentLinkList links={[linkItem]} />,
        );
        expect(
          container
            .getElementsByClassName('listContainer')[0]
            .getElementsByTagName('ul')[0].children,
        ).toHaveLength(0);
      });
    },
  );
});
