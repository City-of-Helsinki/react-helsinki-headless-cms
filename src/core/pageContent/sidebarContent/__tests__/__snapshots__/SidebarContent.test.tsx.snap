// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarContent containing link lists renders a link list when a valid item is included in the links 1`] = `
<div>
  <div
    class="listContainer"
  >
    <h2>
      Renders links into a link list
    </h2>
    <ul
      class="list spacing-s"
    >
      <li>
        <a
          aria-label="link1"
          class="link linkM disableVisitedStyles link link"
          href="#"
        >
          <span
            class="content"
          >
            link1
          </span>
          <span
            aria-hidden="true"
            class="iconRight verticalAlignSmallOrMediumIcon"
          >
            <svg
              aria-hidden="true"
              aria-label="arrow-right"
              class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_small__gTGkU icon_hds-icon--size-s__2Lkik"
              role="img"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M14 5.5L12.5 7L16.5 11H4V13H16.5L12.5 17L14 18.5L20.5 12L14 5.5Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="link2. Opens in a new tab. Opens a different website."
          class="link linkM disableVisitedStyles link link"
          href="https://www.hel.fi"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="content"
          >
            link2
          </span>
          <span
            class="externalWrapper"
          >
            <svg
              aria-hidden="true"
              aria-label="link-external"
              class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC icon verticalAlignSmallOrMediumIcon"
              role="img"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M5 3V19H21V21H3V3H5ZM21 3V12H19V6.413L9.91421 15.5L8.5 14.0858L17.585 5H12V3H21Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
